src/torch_ucc.cpp

```diff
@@ -1009,96 +1080,66 @@ c10::intrusive_ptr<ProcessGroup::Work> ProcessGroupUCC::reduce_scatter(
 	check_tensor(outputTensors);
 	check_device(inputTensors[0][0].device(), outputTensors[0].device());
 	initComm(inputTensors[0][0].device());
-
-	auto isize = inputTensors.size();
   auto data = std::make_unique<WorkData>();
-	std::vector<at::Tensor> flat_input;
-	flat_input.resize(isize);
-	bool isCuda = false;
-#ifdef USE_CUDA
-	isCuda = inputTensors[0][0].device().is_cuda();
-	std::unique_ptr<at::cuda::CUDAEvent> ev = nullptr;
-	if (isCuda) {
-		ev = getPooledEvent();
-		auto sr = at::cuda::getCurrentCUDAStream(
-					inputTensors[0][0].device().index());
-		// mark and wait for inputTensors to be ready
-    ev->record(sr);
-		ev->block(*stream);
-	}
-	// switch to internal stream for copy
-	at::cuda::CUDAStreamGuard guard(*stream);
-#endif
-  for (size_t i = 0; i < isize; i++) {
-		TORCH_CHECK((inputTensors[i].size() == isize * size_),
-			"Tensor input list is not valid for the number of participants");
-		auto onumel = outputTensors[i].numel();
-		flat_input[i] = c10d::newLikeFlat(inputTensors, i);
-		for (size_t j = 0; j < inputTensors[i].size(); j++) {
-    	TORCH_CHECK(
-				(inputTensors[i][j].numel() == onumel),
-				"Tensor operand counts must be same");
-#ifdef USE_CUDA
-			if (isCuda) {
-				c10::cuda::CUDACachingAllocator::recordStream(
-					inputTensors[i][j].storage().data_ptr(), (*stream));
-				flat_input[i][j].copy_(inputTensors[i][j], true);
-			} else
-#endif
-			{
-				flat_input[i][j].copy_(inputTensors[i][j], false);
-			}
-		}
-#ifdef USE_CUDA
-		if (isCuda) {
-			c10::cuda::CUDACachingAllocator::recordStream(
-				outputTensors[i].storage().data_ptr(), (*stream));
-		}
-#endif
+	std::vector<at::Tensor> flat_input(inputTensors.size());
+  for (size_t i = 0; i < inputTensors.size(); i++) {
+    TORCH_CHECK(inputTensors[i].size() == inputTensors.size() * size_,
+      "Tensor input list is not valid for the number of participants");
+      flat_input[i] = c10d::newLikeFlat(inputTensors, i);
   }
+  SAVE_TENSORS(flat_input, data->flat);
 	check_tensor(flat_input);
   ucc_coll_args_t coll;
   coll.mask = UCC_COLL_ARGS_FIELD_PREDEFINED_REDUCTIONS;
   coll.flags = 0;
   coll.coll_type = UCC_COLL_TYPE_REDUCE_SCATTER;
 	coll.reduce.predefined_op = ucc_op_map.at(opts.reduceOp);
-
   coll.src.info.buffer = flat_input[0].data_ptr();
   coll.src.info.count = flat_input[0].numel();
-  coll.src.info.datatype = ucc_dtype_map.at(flat_input[0].scalar_type());
-  coll.src.info.mem_type = to_ucc_memType(flat_input[0].device().type());
+  coll.src.info.datatype = ucc_dtype_map.at(inputTensors[0][0].scalar_type());
+  coll.src.info.mem_type = to_ucc_memType(inputTensors[0][0].device().type());
```

Comment:
minor:
might be get info all from either `inputTensors[0][0]` or `flat_input[0]` to avoid any confusion?